src/main/java/org/vaadin/firitin/components/TreeTable.java

```diff
@@ -0,0 +1,270 @@
+package org.vaadin.firitin.components;
+
+import com.helger.commons.mutable.MutableInt;
+import com.vaadin.flow.component.ClientCallable;
+import com.vaadin.flow.component.Component;
+import com.vaadin.flow.component.HasText;
+import com.vaadin.flow.component.Tag;
+import com.vaadin.flow.component.grid.dataview.GridDataView;
+import com.vaadin.flow.component.grid.dataview.GridLazyDataView;
+import com.vaadin.flow.component.grid.dataview.GridListDataView;
+import com.vaadin.flow.data.provider.BackEndDataProvider;
+import com.vaadin.flow.data.provider.DataProvider;
+import com.vaadin.flow.data.provider.InMemoryDataProvider;
+import com.vaadin.flow.data.provider.ListDataProvider;
+import com.vaadin.flow.function.SerializableFunction;
+import org.vaadin.firitin.components.grid.VGrid;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Plan for this component:
+ * - support lazy loading
+ * - no weird HierarchicalDataProvider stuff
```

Comment:
I'm probably so used to the API by now that I did even think about it as weird anymore :D Only thing that always drives me crazy is that entries have to be the same type (I understand the technical limitation, but still annoying) - so I'm always creating a wrapper class which has information about their "type" (e.g. folder oder file type) and their children so that adding is pretty much setItems(rootItems, Wrapper::getChildren) :)

Reply:
Well, if you are using that version of setItems, you are lucky and avoiding that hierarchicaldataprovider stuff completely. That is the good part of current TreeGrid 😎